mlton/codegen/llvm-codegen/llvm-codegen.fun

```diff
@@ -266,17 +266,88 @@ fun mkgep (lhs, ty, arg, idcs) =
         concat ["\t", lhs, " = getelementptr inbounds ", getTypeFromPointer ty, ", ", ty, " ", arg, ", ", indices, "\n"]
     end
 
+
+
+structure Metadata = struct
+   datatype t =
+      Unnamed of int
+   fun str (Unnamed i) = "!" ^ (Int.toString i)
+
+   val metaDataCounter = ref 0
+   fun new () =
+      let
+         val i = !metaDataCounter
+         val () = Int.inc metaDataCounter
+      in
+         Unnamed i
+      end
+   fun reset () =
+      metaDataCounter := 0
+
+   val rec addSep =
+      fn [] => ""
+       | [s] => s
+       | s :: ss' => concat [s, ", ", addSep ss']
+
+   fun defineNode (t, ts) =
+      concat
+         [str t,
+          " = !{",
+          (addSep o List.map) (ts, str),
+          "}"]
+end
+
+structure SimpleOper = struct
+
+   datatype t = Stack
```

Comment:
While I agree that `Offset` and `SequenceOffset` would not alias, a simple `Stack | Heap | Other` distinction seems simpler.

Reply:
Since I've added the offsets back in, and the complexity seems to have little effect on time for TBAA-based aliasing, I'd argue that there's little benefit to that simplicity.